extensions/amp-analytics/0.1/variables.js

```diff
@@ -309,67 +309,68 @@ export class VariableService {
       }
 
       let value = options.getVar(name);
+      const bindings = opt_bindings || this.getMacros(element);
+      const urlReplacements = Services.urlReplacementsForDoc(element);
 
       if (typeof value == 'string') {
-        value = this.expandValue_(
+        value = this.expandValueAndReplaceAsync_(
           value,
           options,
           element,
+          bindings,
+          urlReplacements,
           opt_bindings,
-          opt_whitelist
+          opt_whitelist,
+          argList
         );
       } else if (isArray(value)) {
         // Treat each value as a template and expand
         for (let i = 0; i < value.length; i++) {
           value[i] =
             typeof value[i] == 'string'
-              ? this.expandValue_(
+              ? this.expandValueAndReplaceAsync_(
                   value[i],
                   options,
                   element,
+                  bindings,
+                  urlReplacements,
                   opt_bindings,
                   opt_whitelist
                 )
               : value[i];
         }
+        value = Promise.all(value);
       }
 
-      const bindings = opt_bindings || this.getMacros(element);
-      const urlReplacements = Services.urlReplacementsForDoc(element);
-
-      return Promise.resolve(value)
-        .then((value) => {
-          if (isArray(value)) {
-            return Promise.all(
-              value.map((item) =>
-                urlReplacements.expandStringAsync(item, bindings, opt_whitelist)
-              )
-            );
-          }
-          return urlReplacements.expandStringAsync(
-            value + argList,
-            bindings,
-            opt_whitelist
-          );
-        })
-        .then((value) => {
-          if (!options.noEncode) {
-            value = encodeVars(/** @type {string|?Array<string>} */ (value));
-          }
-          return value;
-        });
+      return Promise.resolve(value).then((value) =>
+        !options.noEncode
+          ? encodeVars(/** @type {string|?Array<string>} */ (value))
```

Comment:
`#encodeVars` convert the value to string with `String(value)`, do you think we can apply the same to `#expandValueAndReplaceAsync_` as well?

Reply:
If we do:
```
expandValueAndReplaceAsync_(...) {
    return this.expandTemplate(...)
      .then((val) =>
        urlReplacements.expandStringAsync(... )
      )
      .then((val) => String(val));
  }
```

then for variables that resolve to `null`, and `undefined` will turn into `"null"` and `"undefined"`, instead of being empty string. We could account for it like:
```
 ...
 .then((val) => val ? String(val) : '');
  }
```

Which would give us same behavior and be more predictable. Thoughts?

Reply:
I see, looks like we can't switch because it will change the behavior. 
I don't think we can convert the value to string after the promise has resolved either, because `expandAsyncUrl` will be called again in later lifecycle. 
I'm good with what we have now.

Reply:
Thanks for the explanation. 

SGTM